File: leetcodeGo/sliding_window/[blind 75]_BestTimeToBuyAndSellStock.go
func maxProfit(prices []int) int {
    var maxProfit int
    l, r := 0, 1
    for {
        if r >= len(prices) { 
            break 
        }
        if prices[l] < prices[r] {
            currentProfit := prices[r] - prices[l]
            if currentProfit > maxProfit {
                maxProfit = currentProfit
            }
        } else {
            l = r
        }
        r += 1
    }
    return maxProfit    
}

//TC: O(n)
//SC: O(1)

File: leetcode_rust/sliding_window/[blind 75]_BestTimeToBuyAndSellStock.rs
use std::cmp;

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut max_profit = 0;
        let (mut l, mut r) = (0, 0);
        while r < prices.len() {
            if prices[l] < prices[r] { //since i32 implements Copy, there is no need for &prices[idx]
                let mut current_profit = prices[r]-prices[l];
                max_profit = cmp::max(current_profit, max_profit);
            } else {
                l = r; //since i32 implements Copy, there is no need for l = r.clone();
            }
            r += 1;
        }
        max_profit
    }
}

//TC: O(n)
//SC: O(1)
